internal/faults/common.go

```diff
@@ -0,0 +1,278 @@
+		return resumeFn()
+	}
+
+	serverMethod, err := headers.Lookup(ctx, FaultServerMethodHeader)
+	if err != nil {
+		infof("error looking up header %q: %v", FaultServerMethodHeader, err)
+		totalReqsCounter(true, false).Inc()
+		return resumeFn()
+	}
+	if serverMethod != "" && serverMethod != method {
+		infof("method %q does not match the %s header value %q", method, FaultServerMethodHeader, serverMethod)
+		totalReqsCounter(true, false).Inc()
+		return resumeFn()
+	}
+
+	delayMs, err := headers.Lookup(ctx, FaultDelayMsHeader)
+	if err != nil {
+		infof("error looking up header %q: %v", FaultDelayMsHeader, err)
+	}
+	if delayMs != "" {
+		percentageHeader, err := headers.Lookup(ctx, FaultDelayPercentageHeader)
+		if err != nil {
+			infof("error looking up header %q: %v", FaultDelayPercentageHeader, err)
+		}
+		percentage, err := parsePercentage(percentageHeader)
+		if err != nil {
+			warnf("error parsing percentage header %q: %v", FaultDelayPercentageHeader, err)
+			totalReqsCounter(false, false).Inc()
+			return resumeFn()
+		}
+
+		if selected(percentage) {
+			delay, err := strconv.Atoi(delayMs)
+			if err != nil {
+				warnf("unable to convert provided delay %q to integer: %v", delayMs, err)
+				totalReqsCounter(false, false).Inc()
+				return resumeFn()
+			}
+
+			if err := sleep(ctx, time.Duration(delay)*time.Millisecond); err != nil {
+				warnf("error when delaying request: %v", err)
+				totalReqsCounter(false, false).Inc()
+				return resumeFn()
+			}
+			delayed = true
+		}
+	}
+
+	abortCode, err := headers.Lookup(ctx, FaultAbortCodeHeader)
+	if err != nil {
+		infof("error looking up header %q: %v", FaultAbortCodeHeader, err)
+	}
+	if abortCode != "" {
+		percentageHeader, err := headers.Lookup(ctx, FaultAbortPercentageHeader)
+		if err != nil {
+			infof("error looking up header %q: %v", FaultAbortPercentageHeader, err)
+		}
+		percentage, err := parsePercentage(percentageHeader)
+		if err != nil {
+			warnf("error parsing percentage header %q: %v", FaultAbortPercentageHeader, err)
+			totalReqsCounter(false, false).Inc()
+			return resumeFn()
+		}
+
+		if selected(percentage) {
+			code, err := strconv.Atoi(abortCode)
+			if err != nil {
+				warnf("unable to convert provided abort %q to integer: %v", abortCode, err)
+				totalReqsCounter(false, false).Inc()
+				return resumeFn()
+			}
+			if code < i.abortCodeMin || code > i.abortCodeMax {
+				warnf("provided abort code \"%d\" is outside of the valid range [%d-%d]", code, i.abortCodeMin, i.abortCodeMax)
+				totalReqsCounter(false, false).Inc()
+				return resumeFn()
+			}
+
+			abortMessage, err := headers.Lookup(ctx, FaultAbortMessageHeader)
+			if err != nil {
+				warnf("error looking up header %q: %v", FaultAbortMessageHeader, err)
+				totalReqsCounter(false, false).Inc()
+				return resumeFn()
+			}
+
+			totalReqsCounter(true, true).Inc()
+			return faultFn(code, abortMessage)
+		}
+	}
+
+	totalReqsCounter(true, false).Inc()
+	return resumeFn()
+}
```

Comment:
This function appears to be too long to easily glance over it. Can portions of it be factored into helper functions instead?